### ISSUE TITLE
[deeplab]Performance model in cityscape model?

### ISSUE THREAD

**Issue by yjl9122** [tensorflow/models#4313]
i test your model on test data. Then i fineturn your model on train fine+val. I don't get 82.1.

python deeplab/train.py \
    --logtostderr \
    --training_number_of_steps=90000 \
    --train_split="trainval" \
    --model_variant="xception_65" \
    --atrous_rates=6 \
    --atrous_rates=12 \
    --atrous_rates=18 \
    --output_stride=16 \
    --decoder_output_stride=4 \
    --train_crop_size=769 \
    --train_crop_size=769 \
    --train_batch_size=16 \
    --dataset="cityscapes" \
    --tf_initial_checkpoint=${PATH_TO_INITIAL_CHECKPOINT} \
    --train_logdir=${PATH_TO_TRAIN_DIR} \
    --dataset_dir=${PATH_TO_DATASET} \
    --num_clones=4 \
    --base_learning_rate=0.001 \
    --fine_tune_batch_norm=True

python deeplab/vis.py \
    --logtostderr \
    --vis_split="test" \
    --model_variant="xception_65" \
    --atrous_rates=6 \
    --atrous_rates=12 \
    --atrous_rates=18 \
    --output_stride=16 \
    --decoder_output_stride=4 \
    --vis_crop_size=1025 \
    --vis_crop_size=2049 \
    --dataset="cityscapes" \
    --colormap_type="cityscapes" \
    --checkpoint_dir=${PATH_TO_CHECKPOINT} \
    --vis_logdir=${PATH_TO_VIS_DIR} \
    --dataset_dir=${PATH_TO_DATASET} \
    --eval_scales=0.5 \
    --eval_scales=0.75 \
    --eval_scales=1.0 \
    --eval_scales=1.25 \
    --eval_scales=1.5 \
    --eval_scales=1.75 \
    --add_flipped_images=True \
    --also_save_raw_predictions=True

Metric | Value
-- | --
IoU Classes       | 79.536
iIoU Classes       | 56.955
IoU Categories  | 91.1075
iIoU Categories | 78.995

Class | IoU | iIoU
-- | -- | --
road           | 98.7174 | -
sidewalk     | 87.0757 | -
building      | 93.2443 | -
wall             | 54.228 | -
fence           | 59.4212 | -
pole             | 67.6204 | -
traffic light   | 74.3525 | -
traffic sign    | 79.4047 | -
vegetation    | 93.6022 | -
terrain          | 72.3044 | -
sky               | 95.8473 | -
person          | 86.6916    | 68.756
rider              | 71.5668    | 48.2658
car                 | 95.8984    | 90.1608
truck              | 71.046      | 39.045
bus                 | 83.1989   | 51.1338
train               | 80.3375    | 50.3262
motorcycle     | 69.8121    | 46.3885
bicycle            | 76.8155    | 61.5635

Category | IoU | iIoU
-- | -- | --
flat | 98.7293 | -
nature | 93.3203 | -
object | 73.9022 | -
sky | 95.8473 | -
construction | 93.6702 | -
human | 86.8623 | 69.6746
vehicle | 95.4207 | 88.3153

**Comment by yjl9122** [comment]
@aquariusjay Would you give me some suggestions?

**Comment by aquariusjay** [comment]
Hi yjl9122,

Thanks for the inquiry.
The model uses output_stride = 8 and has been further trained on the coarse annotations. We will provide that checkpoint in the near future.

**Comment by yjl9122** [comment]
@aquariusjay How much batch size on each GPU do you set? And it matters on cityscapes?

**Comment by aquariusjay** [comment]
We use batch size 8 with crop size = 769x769, and output_stride = 16 on Cityscapes.
Training with Batch norm is essential to attain high performance.

**Comment by yjl9122** [comment]
@aquariusjay how to use coarse and fine dataset on each stage? and how to set learning rate on each stage?

**Comment by zhoufengbuaa** [comment]
@aquariusjay So it is very essential to sync BN when training your code? If not, 769 * 769 crop size is too big, and each GPU can only contain 2 batches, and it is very harmful for image segmentation task. Am I right??

**Comment by zhoufengbuaa** [comment]
@yjl9122 Using there code and released model, I got a 80.9 IOU on cityscapes validation set. However, I cannot retrain their code. I do not know why. And you??

**Comment by yjl9122** [comment]
@zhoufengbuaa I got same result with author on val, but i don't got same result on test dataset.

**Comment by zhoufengbuaa** [comment]
@yjl9122 Did you retrain the code? I found it very difficult to reproduce the performance shown in the paper， even I used 8 GPU cards.  It takes very long time to train and the loss cannot be converged.

**Comment by yjl9122** [comment]
@zhoufengbuaa segmentation is very hard task, so it is normal.

**Comment by zhoufengbuaa** [comment]
@yjl9122 Yes, it is very, however we already code and the performance is so ok. The problem is we cannot reproduce the performance shown in the paper using his (author) own code. That doesn't make sense.

**Comment by yjl9122** [comment]
@aquariusjay  Could you tell me how to use coarse and fine dataset on each stage? and how to set learning rate on each stage?

**Comment by tt-leader** [comment]
@yjl9122 I think we can use it by simply  adding coarse annotations data to train+val data. The model trains only in one stage and follows the training protocols in paper deeplab v3

**Comment by baoruxiao** [comment]
Any one know the official base learning for training deeplabv3+ on cityscapes? 0.007 or 0.001 as shown in the above configuration? Thanks

**Comment by baoruxiao** [comment]
Hi @yjl9122 

Did you fine-tune from xception 65 imagenet pretrained weights or from provided deeplabv3+ weights.

Thanks

**Comment by tonyli0803** [comment]
@aquariusjay @jianlong-yuan

I try to use 2 Nvidia 1080Ti to reproduce ( retrain )  the deeplab model on cityscape dataset.

However, the gpu memory is not enough to run the code.

I consider to reproduce ( retrain )  the deeplab model on google cloud platform.
But I wonder how good the hardware need to be to run such a large batch size (batch size).

What hardware did you use to run the code?

**Comment by sanmudaxia** [comment]
@jianlong-yuan Hi,I enter the same question of you,  did you solve it and get a better results? and how to change the eval.py so i can get per class IOU like yours ? Thanks for your help!

**Comment by shadowwalkers** [comment]
@jianlong-yuan , hi, you were able to reproduce author's result on val set. Did you use other pretraining  (other than imagenet, like coco, mapillary, JFT) ? Also, did you use coarse data as well for getting comparable val results. 
Thanks
